PYME/IO/DataSources/BioformatsDataSource.py

```diff
@@ -58,44 +62,99 @@ def release_VM():
 
 from .BaseDataSource import BaseDataSource
 
+
+class BioformatsFile(bioformats.ImageReader):
+    def __init__(self, path=None, url=None, perform_init=True):
+        # self.path is always the file path or the url, whichever is specified
+        # url overrides a file path 
+        ensure_VM()
+        self._md = None
+        self._series_count = None
+        self._series_names = None
+        super(BioformatsFile, self).__init__(path, url, perform_init)
+
+    @property
+    def series_count(self):
+        if self._series_count is None:
+            self._series_count = self.rdr.getSeriesCount()
+        return self._series_count
+
+    @property
+    def md(self):
+        if self._md is None:
+            self._md = []
+            _io = self.rdr.getSeries()
+            idxs = np.arange(self.series_count)
+            for _i in idxs:
+                self.rdr.setSeries(_i)
+                series_md = javabridge.jutil.jdictionary_to_string_dictionary(self.rdr.getSeriesMetadata())
+                self._md.append(series_md)
+            self.rdr.setSeries(_io)
+        return self._md
+
+    @property
+    def series_names(self):
+        # return a list of series names in the file
+        if self._series_names is None:
+            self._series_names = []
+            idxs = np.arange(self.series_count)
+            try:
+                for _i in idxs:
+                    self._series_names.append(self.md[_i]['Image name'])
+            except(KeyError):
+                print('Image names not found, naming by integers.')
+                self._series_names = ['Image {}'.format(_i) for _i in idxs.astype(str)]
+        return self._series_names
+
+    def __del__(self):
+        release_VM()
+
 class DataSource(BaseDataSource):
     moduleName = 'BioformatsDataSource'
-    def __init__(self, filename, taskQueue=None, chanNum = 0):
-        self.filename = getFullExistingFilename(filename)#convert relative path to full path
+    def __init__(self, path=None, url=None, taskQueue=None, chanNum = 0, series=None):
```

Comment:
Can't change the positional argument signature (before `chanNum` - can add things after `chanNum`) here, due to the way datasources are passed around in tasks. 

What is the `url` argument supposed to do? I'm guessing this is to allow it to be passed to `bioformats.ImageReader`, but seeing as it's completely undocumented I don't think we need/want to support it. If we do want to support it, I'd suggest we infer whether our filename is a path or a url before passing it.

Reply:
Bioformats *should* do this inference internally - it's a bit stink that it doesn't

Reply:
Yeah, I just copied how bioformats did it since I assumed there was a reason. No reason we can't infer a URL, though.